### PR TITLE
[REF] (purch_)stock,mrp_subcon,dropship: move replenish route to mixin

### PR DESCRIPTION
**Pull request by Sehnde**
The rework on mrp_mps needs to use the same logic for route_id as in the replenish wizard. 
Instead of duplicating the code and creating multiple submodules, we move the route_id logic to a mixin.

see odoo/enterprise#56128

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/purchase_stock/models/stock_replenish_mixin.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    supplier_id = fields.Many2one('product.supplierinfo', string="Vendor")
+    show_vendor = fields.Boolean(compute='_compute_show_vendor')
+
+    @api.depends('route_id')
+    def _compute_show_vendor(self):
+        for rec in self:
+            rec.show_vendor = rec._get_show_vendor(rec.route_id)
+
+    def _get_show_vendor(self, route):
+        return route == self.env.ref('purchase_stock.route_warehouse0_buy', raise_if_not_found=False)
```
Comment (amoyaux) [inline addons/purchase_stock/models/stock_replenish_mixin.py:18]: should remove the ref to xmlid

File: addons/stock/models/stock_replenish_mixin.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+class ProductReplenishMixin(models.AbstractModel):
+    _name = 'stock.replenish.mixin'
+    _description = 'Product Replenish Mixin'
+
+    route_id = fields.Many2one(
+        'stock.route', string="Preferred Route",
+        help="Apply specific route for the replenishment instead of product's default routes.",
+        check_company=True)
+    allowed_route_ids = fields.Many2many('stock.route', compute='_compute_allowed_route_ids')
```
Comment (amoyaux) [inline addons/stock/models/stock_replenish_mixin.py:14]: to clean if a new PR. Remove dynamic domain

File: addons/mrp/models/stock_replenish_mixin.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    bom_id = fields.Many2one('mrp.bom', string="Bill of Material")
+    show_bom = fields.Boolean(compute='_compute_show_bom')
+
+    @api.depends('route_id')
+    def _compute_show_bom(self):
+        for rec in self:
+            rec.show_bom = rec._get_show_bom(rec.route_id)
+
+    def _get_show_bom(self, route):
+        return route == self.env.ref('mrp.route_warehouse0_manufacture', raise_if_not_found=False)
```
[the remark this answers is not in the document]
Comment (Sehnde) [reply addons/mrp/models/stock_replenish_mixin.py]: done

File: addons/purchase_stock/models/stock_replenish_mixin.py (hunk 1 above)
Comment (Sehnde) [reply addons/purchase_stock/models/stock_replenish_mixin.py]: done

File: addons/stock_dropshipping/models/__init__.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
+from . import stock_replenish_mixin
```
[the remark this answers is not in the document]
Comment (Sehnde) [reply addons/stock_dropshipping/models/__init__.py]: done, it's because it was prodict_replenish_mixin before and I forgot to move the line

**Comment by amoyaux** [comment]
robodoo r+ rebase-ff

**Pull request closed without merging**